src/Lecture2.hs

```diff
@@ -1,170 +1,268 @@
+--
+-- >>> lazyProduct [4, 3, 7]
+-- 84
 lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
-
-{- | Implement a function that duplicates every element in the list.
-
->>> duplicate [3, 1, 2]
-[3,3,1,1,2,2]
->>> duplicate "cab"
-"ccaabb"
--}
+lazyProduct lst = case lst of
+  [] -> 1
+  (0 : _) -> 0
+  [x] -> x
+  (x : xs) -> x * lazyProduct xs
+
+-- | Implement a function that duplicates every element in the list.
+--
+-- >>> duplicate [3, 1, 2]
+-- [3,3,1,1,2,2]
+-- >>> duplicate "cab"
+-- "ccaabb"
 duplicate :: [a] -> [a]
-duplicate = error "TODO"
-
-{- | Implement function that takes index and a list and removes the
-element at the given position. Additionally, this function should also
-return the removed element.
-
->>> removeAt 0 [1 .. 5]
-(Just 1,[2,3,4,5])
-
->>> removeAt 10 [1 .. 5]
-(Nothing,[1,2,3,4,5])
--}
-removeAt = error "TODO"
-
-{- | Write a function that takes a list of lists and returns only
-lists of even lengths.
-
->>> evenLists [[1,2,3], [3,1,2,7], [], [5, 7, 2]]
-[[3,1,2,7],[]]
-
-♫ NOTE: Use eta-reduction and function composition (the dot (.) operator)
-  in this function.
--}
-evenLists = error "TODO"
+duplicate lst = case lst of
+  [] -> []
+  (x : xs) -> x : x : duplicate xs
+
+-- | Implement function that takes index and a list and removes the
+-- element at the given position. Additionally, this function should also
+-- return the removed element.
+--
+-- >>> removeAt 0 [1 .. 5]
+-- (Just 1,[2,3,4,5])
+--
+-- >>> removeAt 10 [1 .. 5]
+-- (Nothing,[1,2,3,4,5])
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt = removeHelper []
+  where
+    -- cannot use length, lst can be infinite
+    removeHelper :: [a] -> Int -> [a] -> (Maybe a, [a])
+    removeHelper before index after
+      | index < 0 = (Nothing, after)
+      | null after = (Nothing, before)
+      | length before == index = (Just (head after), before ++ drop 1 after)
+      | otherwise = removeHelper (before ++ take 1 after) index (drop 1 after)
+
+-- | Write a function that takes a list of lists and returns only
+-- lists of even lengths.
+--
+-- >>> evenLists [[1,2,3], [3,1,2,7], [], [5, 7, 2]]
+-- [[3,1,2,7],[]]
+--
+-- ♫ NOTE: Use eta-reduction and function composition (the dot (.) operator)
+--  in this function.
+evenLists :: [[a]] -> [[a]]
+evenLists = filter (even . length)
+
+-- | The @dropSpaces@ function takes a string containing a single word
+-- or number surrounded by spaces and removes all leading and trailing
+-- spaces.
+--
+-- >>> dropSpaces "   hello  "
+-- "hello"
+-- >>> dropSpaces "-200            "
+-- "-200"
+--
+-- ♫ NOTE: As in the previous task, use eta-reduction and function
+--  composition (the dot (.) operator) in this function.
+--
+-- 🕯 HINT: look into Data.Char and Prelude modules for functions you may use.
+dropSpacesBefore :: String -> String
+dropSpacesBefore str
+  | null str = str
+  | isSpace (head str) = dropSpacesBefore (tail str)
+  | otherwise = str
```

Comment:
That's the correct solution 👍🏻 
You can also use the standard function [dropWhile](https://hackage.haskell.org/package/base-4.17.0.0/docs/Prelude.html#v:dropWhile) to implement `dropSpacesBefore` shorter:

```suggestion
dropSpacesBefore = dropWhile isSpace
```